### PR TITLE
Delayed transactions

### LINKED ISSUE #3594: TransactionQueue should support a minimum block number before propagation/inclusion
`parity_postTransaction` should be able to accept a block number before which the transaction stays in the queue but is not propagated or enacted. this will allow for users and their dapps to schedule and sign transactions in advance.

### PR DESCRIPTION
**Pull request by arkpar**
Closes #3594

### CODE REVIEW

File: rpc/src/v1/types/confirmations.rs [hunk 1]
```diff
@@ -324,7 +329,8 @@ mod tests {
 	fn should_deserialize_modification() {
 		// given
 		let s1 = r#"{
-			"gasPrice":"0xba43b7400"
+			"gasPrice":"0xba43b7400",
+			"minBlock":42
```
Comment (tomusdrw) [inline rpc/src/v1/types/confirmations.rs]: I think we encode block numbers as hex almost everywhere, would be good to keep consistency.

File: rpc/src/v1/types/transaction_request.rs [hunk 2]
```diff
@@ -41,6 +41,9 @@ pub struct TransactionRequest {
 	pub data: Option<Bytes>,
 	/// Transaction's nonce
 	pub nonce: Option<U256>,
+	/// Delay until this block if specified.
+	#[serde(rename="minBlock")]
+	pub min_block: Option<u64>,
```
Comment (tomusdrw) [inline rpc/src/v1/types/transaction_request.rs]: Consistency for `QUANTITY` encoding for block number.

File: ethcore/src/miner/transaction_queue.rs [hunk 3]
```diff
@@ -968,27 +974,56 @@ impl TransactionQueue {
...
-	/// Returns local transactions (some of them might not be part of the queue anymore).
-	pub fn local_transactions(&self) -> &LinkedHashMap<H256, LocalTransactionStatus> {
-		self.local_transactions.all_transactions()
+	fn collect_pending_transaction<F>(&self, best_block: BlockNumber, mut f: F)
+		where F: FnMut(&VerifiedTransaction) {
+
+		let mut delayed = HashSet::new();
+		for t in self.current.by_priority.iter() {
+			let tx = self.by_hash.get(&t.hash).expect("All transactions in `current` and `future` are always included in `by_hash`");
+			let sender = tx.transaction.sender().expect("Queue only contains transactions with valid sender");
```
Comment (tomusdrw) [inline ethcore/src/miner/transaction_queue.rs]: Just use `tx.sender()` it doesn't require the proof.

File: ethcore/src/miner/transaction_queue.rs [hunk 4]
```diff
@@ -968,27 +974,56 @@ impl TransactionQueue {
...
-			.map(|t| t.transaction.clone())
-			.collect()
+		self.top_transactions_at(BlockNumber::max_value())
+
 	}
 
-	/// Returns local transactions (some of them might not be part of the queue anymore).
-	pub fn local_transactions(&self) -> &LinkedHashMap<H256, LocalTransactionStatus> {
-		self.local_transactions.all_transactions()
+	fn collect_pending_transaction<F>(&self, best_block: BlockNumber, mut f: F)
```
Comment (tomusdrw) [inline ethcore/src/miner/transaction_queue.rs]: Wouldn't it be better to `filter_pending_transactions`?

File: ethcore/src/types/transaction.rs [hunk 5]
```diff
@@ -390,6 +390,34 @@ impl Deref for LocalizedTransaction {
 	}
 }
 
+/// Queued information with additional information.
```
Comment (tomusdrw) [inline ethcore/src/types/transaction.rs]: "Queued information" -> "Queued transaction"

File: ethcore/src/types/transaction.rs [hunk 6]
```diff
@@ -390,6 +390,34 @@ impl Deref for LocalizedTransaction {
...
+#[derive(Debug, Clone, PartialEq, Eq, Binary)]
+pub struct PendingTransaction {
```
Comment (tomusdrw) [inline ethcore/src/types/transaction.rs]: You could implement `Deref<Target=SignedTransaction>` to get rid of some `tx.transaction.<functioncall>` statements.

Comment (arkpar) [inline ethcore/src/types/transaction.rs]: I'd prefer it to be explicit. Makes the code easier to follow

File: ethcore/src/types/transaction.rs [hunk 7]
```diff
@@ -390,6 +390,34 @@ impl Deref for LocalizedTransaction {
 	}
 }
 
+/// Queued transaction with additional information.
+#[derive(Debug, Clone, PartialEq, Eq, Binary)]
+pub struct PendingTransaction {
+	/// Signed transaction data.
+	pub transaction: SignedTransaction,
+	/// Gas price.
+	pub min_block: Option<BlockNumber>,
```
Comment (rphmeier) [inline ethcore/src/types/transaction.rs]: comment is wrong

File: sync/src/chain.rs [hunk 8]
```diff
@@ -1917,10 +1917,10 @@ impl ChainSync {
 			return 0;
 		}
 
-		let all_transactions_hashes = transactions.iter().map(|tx| tx.hash()).collect::<HashSet<H256>>();
+		let all_transactions_hashes = transactions.iter().map(|tx| tx.transaction.hash()).collect::<HashSet<H256>>();
```
Comment (gavofyork) [inline sync/src/chain.rs]: does this send out pending transactions, too?

File: ethcore/src/miner/mod.rs [hunk 9]
```diff
@@ -144,11 +144,14 @@ pub trait MinerService : Send + Sync {
 	/// Query pending transactions for hash.
 	fn transaction(&self, best_block: BlockNumber, hash: &H256) -> Option<SignedTransaction>;
 
-	/// Get a list of all transactions.
-	fn all_transactions(&self) -> Vec<SignedTransaction>;
+	/// Get a list of all ready transactions in the queue.
+	fn all_transactions(&self) -> Vec<PendingTransaction>;
```
Comment (gavofyork) [inline ethcore/src/miner/mod.rs]: rename to `ready_transactions` to make the intent clear?

**Comment by gavofyork** [comment]
looks ok, assuming that the pending transactions are indeed not sent out onto the network. would be nice to have one or more test demonstrating the functionality though...

File: sync/src/chain.rs (hunk 8 above)
Comment (arkpar) [inline sync/src/chain.rs]: only those that can go into the next block

**Pull request merged**